rfc/api.md

```diff
@@ -0,0 +1,120 @@
+
+# Summary
+
+Stabilize Framework API
+
+# Motivation
+
+To prevent spontaneous API changes just because someone wants it.
+
+# Detailed design
+
+The lowest level class is the connection (may be renamed to session). It does almost all low level network logic. It's very likely that it will be hidden from the userscope.
+
+The main class you will deal with - is a service. It represents a proxy with auto resolving and RAII style.
+
+It's very unlikely that you will create service class instances manually. Some entity called service manager should be designed to create it for you.
+
+RAII style means that its destructor will block until all internal pending callbacks are called, which results in asynchronous invocation of all futures that you have gotten (via `then`).
+
+A service can be detached, then its destructor won't block.
+
+All you need to create a service class instance via service manager it its name and prototype. 
+
+```
+auto node = manager.create<T>("node");
+```
+
+The obtained service is initially in disconnected state. To resolve and connect manually you need to call `connect()` method, which returns flagged future. You can always check is the service is in connected state by calling `connected()` method. Note, that this method does passive connection checking, and it can result in false-positive value if the real connection is already broken, but there are no internal callbacks attached for reading or writing at that time.
+
+The service supports auto reconnection on new requests.
+
+To invoke remote method you need to call `invoke<Event>(args...)` method with appropriate event protocol description and its parameters. This one returns you a future of pair of sender/receiver, by manipulating with each of them you can control an upstream and downstream.
+
+```
+template<class T>
+class service {
+    ~service();
+  
+    void detach();
+  
+    auto connected() const noexcept -> bool;
+    auto connect() -> future<void>;
+
+    template<class Event, class... Args>
+    auto invoke(Args&&... args) -> future<std::tuple<sender<T>, receiver<T>>>;
+};
+```
+
+Cocaine upstream and downstream are completely separated entities. In the common case when you send another event in the open channel you make one step forward your protocol transition graph. There are terminate leafs, which means you cannot send another event when you reach it. On the receiver side it is completely identical image - there are receiver state, after reaching which you cannot receive more chunks.
+
+This complex graph structure requires more runtime checks which can be theoretically replaced with static checks, which gives us the following API.
+
+Transmission channel does step forward after invoking `send` method, which accepts event type and its arguments described in protocol `dispatch_type`. It returns the following transmission state, which can be reassigned to the old `tx`, because it's useless now, which gives us an ability to chain requests:
```

Comment:
> which can be reassigned to the old `tx`

As I understand, tx's type depends on the transition graph state, so this statement is not true in the general case, as well as the code snippets below.